### PR TITLE
An in-memory RNG that shares its file descriptor.

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR implements an in-memory random number generator that only uses an OS RNG for (re)seeding. The OS RNG is shared, so there's only one file descriptor for `/dev/urandom` being used.

The PR also implements a tidy check that we don't accidentally introduce an RNG. Rather annoyingly, there are a lot of transitive dependencies on `rand`, notably hash maps in `std`.

This PR makes it possible to use uuids for identifiers such as pipeline and frame ids.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because it's fixing a resource issue

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14351)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @ConnorGBrewster @larsbergstrom @nox

**Comment by asajeffrey** [comment]
@avadacatavra this replaces the RNG in crypto. We should double-check that we are okay with that.

**Comment by emilio** [comment]
FWIW, rayon uses `rand` but only for `XorShiftRng`, so probably we should be able to filter this in a better way?

**Comment by asajeffrey** [comment]
We don't scan dependency source, so it's difficult to see how we can do much better than a whitelist.

**Comment by nox** [comment]
That's quite a lot of code. I would rather we use the `Fnv` one explicitly when we see failures due to file limits with a RNG-related stack, than use a custom hasher.

**Comment by asajeffrey** [comment]
@nox: yes, we can use Fnv for hash maps, but that doesn't help for uuids or other cases where we need an RNG, not a hash function.

Most of the code here is just cut-and-paste from `rand`, in particular there's no new RNG algorithms, it just reuses existing ones. The only new bit is sharing an `OsRng` (and hence sharing an fd) rather than creating a new one for each thread.

**Comment by asajeffrey** [comment]
Rebased.

**Comment by asajeffrey** [comment]
Rebased. @emilio or @nox: any chance of a review?

**Comment by emilio** [comment]
This is fine for me when rebased, but maybe @nox has still concerns.

File: python/tidy/servo_tidy/tidy.py [hunk 1]
```diff
@@ -81,6 +81,24 @@
     " accessible to\n// web pages."
 ]
 
+WHITELISTED_DEPENDENCIES = {
```
Comment (Wafflespeanut) [inline python/tidy/servo_tidy/tidy.py]: We should really be moving this to the config file, but that shouldn't be a concern for this PR. I'll take care of it in an easy issue :smile:

**Comment by asajeffrey** [comment]
IRC conversation with @nox: http://logs.glob.uno/?c=mozilla%23servo&s=5+Jan+2017&e=5+Jan+2017#c587629

TL;DR he r+s once it's split into two commits, and `WHITELISTED_DEPENDENCIES` is renamed.
